homeassistant/components/homekit/type_locks.py

```diff
@@ -0,0 +1,86 @@
+    CATEGORY_LOCK, SERV_LOCK,
+    CHAR_LOCK_CURRENT_STATE, CHAR_LOCK_TARGET_STATE)
+
+_LOGGER = logging.getLogger(__name__)
+
+HASS_TO_HOMEKIT = {STATE_UNLOCKED: 0,
+                   STATE_LOCKED: 1,
+                   # value 2 is Jammed which hass doesn't have a state for
+                   STATE_UNKNOWN: 3}
+HOMEKIT_TO_HASS = {c: s for s, c in HASS_TO_HOMEKIT.items()}
+STATE_TO_SERVICE = {STATE_LOCKED: 'lock',
+                    STATE_UNLOCKED: 'unlock'}
+
+
+@TYPES.register('Lock')
+class Lock(HomeAccessory):
+    """Generate a Lock accessory for a lock entity.
+
+    The lock entity must support: unlock and lock.
+    """
+
+    def __init__(self, hass, entity_id, name, *args, **kwargs):
+        """Initialize a Lock accessory object."""
+        super().__init__(name, entity_id, CATEGORY_LOCK, *args, **kwargs)
+
+        self._hass = hass
+        self._entity_id = entity_id
+
+        self.flag_target_state = False
+
+        serv_lock_mechanism = add_preload_service(self, SERV_LOCK)
+        self.char_current_state = serv_lock_mechanism. \
+            get_characteristic(CHAR_LOCK_CURRENT_STATE)
+        self.char_target_state = serv_lock_mechanism. \
+            get_characteristic(CHAR_LOCK_TARGET_STATE)
+
+        self.char_current_state.value = HASS_TO_HOMEKIT[STATE_UNKNOWN]
+        self.char_target_state.value = HASS_TO_HOMEKIT[STATE_LOCKED]
+
+        self.char_target_state.setter_callback = self.set_state
+
+    def set_state(self, value):
+        """Set lock state to value if call came from HomeKit."""
+        _LOGGER.debug("%s: Set state to %d", self._entity_id, value)
+        self.flag_target_state = True
+        self.char_target_state.set_value(value, should_callback=False)
+
+        hass_value = HOMEKIT_TO_HASS.get(value)
+        service = STATE_TO_SERVICE[hass_value]
+
+        params = {ATTR_ENTITY_ID: self._entity_id}
+        self._hass.services.call('lock', service, params)
+
+    def update_state(self, entity_id=None, old_state=None, new_state=None):
+        """Update lock after state changed."""
+        if new_state is None:
+            return
+
+        hass_state = new_state.state
+        if HASS_TO_HOMEKIT.get(hass_state) is None:
+            _LOGGER.debug('%s: state %s is not supported',
+                          entity_id,
+                          hass_state)
+            return
+
+        current_lock_state = HASS_TO_HOMEKIT[hass_state]
+        self.char_current_state.set_value(current_lock_state,
+                                          should_callback=False)
+        _LOGGER.debug('%s: Updated current state to %s (%d)',
+                      self._entity_id, hass_state, current_lock_state)
+
+        if not self.flag_target_state:
+            self.char_target_state.set_value(current_lock_state,
+                                             should_callback=False)
+        if self.char_target_state.value == self.char_current_state.value:
+            self.flag_target_state = False
```

Comment:
L71-73: Nobody will see the debug statement.
L85: Shouldn't be necessary.

Maybe that should work:
```py
        hass_state = new_state.state
        if HASS_TO_HOMEKIT.get(hass_state) is not None:
            current_lock_state = HASS_TO_HOMEKIT[hass_state]
            self.char_current_state.set_value(current_lock_state,
                                              should_callback=False)
            _LOGGER.debug('%s: Updated current state to %s (%d)',
                          self._entity_id, hass_state, current_lock_state)
            if not self.flag_target_state:
                self.char_target_state.set_value(current_lock_state,
                                                 should_callback=False)
            self.flag_target_state = False
```

Reply:
I have that check in case additional `STATE_*` are added to the lock component in the future, that seems like a reasonable possibility to me? (`STATE_JAMMED` for instance).

Right on L85, curious why `security_system` does it now that I think about it. Taking a look at that (and turning up logs in test) did lead me to noticing an exception (that for some reason doesn't cause tests to fail even though they cause it?):

```
ERROR:homeassistant.core:Error doing job: Future exception was never retrieved
Traceback (most recent call last):
  File "/Users/pkates/.pyenv/versions/3.6.4/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/pkates/pksrc/home-assistant/homeassistant/components/homekit/type_locks.py", line 84, in update_state
    should_callback=False)
  File "/Users/pkates/.pyenv/versions/3.6.4/envs/hass/lib/python3.6/site-packages/pyhap/characteristic.py", line 150, in set_value
    raise ValueError
ValueError
```

This happens when Hass sets STATE_UNKNOWN which target_state doesn't support as a valid value.

My first thought was just `if state == STATE_UNLOCKED` but that feels lazy and brittle, I went with a try/except but I definitely willing to hear alternate opinions.

Reply:
Maybe try: `if hass_state in (STATE_UNLOCKED, STATE_LOCKED)`. I wouldn't agree that it's lazy, just practical and easier to read.

You don't need any debug statements for state not supported. That would get out of hand otherwise.